src/chunker/chunker.js

```diff
@@ -332,48 +333,63 @@ class Chunker {
 
 	async *layout(content, startAt) {
 		let breakToken = startAt || false;
-		let tokens = [];
+		let page, prevPage, prevNumPages;
 
 		while (breakToken !== undefined && (MAX_PAGES ? this.total < MAX_PAGES : true)) {
 
-			if (breakToken && breakToken.node) {
-				await this.handleBreaks(breakToken.node);
-			} else {
-				await this.handleBreaks(content.firstChild);
+			let addedExtra = false;
+			let emptyBody = !page || !page.area.firstElementChild.childElementCount || !page.area.firstElementChild.firstElementChild.getBoundingClientRect().height;
+			let emptyFootnotes = !page || !page.footnotesArea.firstElementChild.childElementCount || !page.footnotesArea.firstElementChild.firstElementChild.getBoundingClientRect().height;
+			let emptyPage = (emptyBody && emptyFootnotes);
+
+			prevNumPages = this.total;
+
+			if (!page || !emptyPage) {
+				if (breakToken) {
+					if (breakToken.overflow.length && breakToken.overflow[0].node) {
+						// Overflow.
+						await this.handleBreaks(breakToken.overflow[0].node);
+					}
+					else {
+						await this.handleBreaks(breakToken.node);
+					}
+				} else {
+					await this.handleBreaks(content.firstChild);
+				}
+			}
+
+			addedExtra = this.total != prevNumPages;
+
+			// Don't add a page if we have a forced break now and we just
+			// did a break due to overflow but have nothing displayed on
+			// the current page, unless there's overflow and we're finished.
+			if (!page || addedExtra || !emptyPage) {
+				this.addPage();
 			}
 
-			let page = this.addPage();
+			page = this.pages[this.total - 1];
 
 			await this.hooks.beforePageLayout.trigger(page, content, breakToken, this);
 			this.emit("page", page);
 
 			// Layout content in the page, starting from the breakToken
-			breakToken = await page.layout(content, breakToken, this.maxChars);
-
 			if (breakToken) {
-				let newToken = breakToken.toJSON(true);
-				if (tokens.lastIndexOf(newToken) > -1) {
-					// loop
-					let err = new OverflowContentError("Layout repeated", [breakToken.node]);
-					console.error("Layout repeated at: ", breakToken.node);
-					return err;
-				} else {
-					tokens.push(newToken);
-				}
+				// Debugging tool - ensures previously rendered content is visible.
+				window.scrollTo(0, document.body.scrollHeight);
```

Comment:
If this is just for debugging, I think we should either remove this or put it behind a debug flag. 

When testing with the Moby Dick file, I'm seeing ~200ms faster rendering without this scrolling.

Reply:
Ok, I'll remove.